pinot-segment-local/src/main/java/org/apache/pinot/segment/local/dedup/BasePartitionDedupMetadataManager.java

```diff
@@ -0,0 +1,241 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.dedup;
+
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import org.apache.pinot.common.metrics.ServerGauge;
+import org.apache.pinot.common.metrics.ServerMetrics;
+import org.apache.pinot.common.metrics.ServerTimer;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.spi.config.table.HashFunction;
+import org.apache.pinot.spi.data.readers.PrimaryKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public abstract class BasePartitionDedupMetadataManager implements PartitionDedupMetadataManager {
+  protected final String _tableNameWithType;
+  protected final List<String> _primaryKeyColumns;
+  protected final int _partitionId;
+  protected final ServerMetrics _serverMetrics;
+  protected final HashFunction _hashFunction;
+  protected final double _metadataTTL;
+  protected final String _dedupTimeColumn;
+  protected final Logger _logger;
+
+  // The following variables are always accessed within synchronized block
+  private boolean _stopped;
+  // Initialize with 1 pending operation to indicate the metadata manager can take more operations
+  private int _numPendingOperations = 1;
+  private boolean _closed;
+
+  protected BasePartitionDedupMetadataManager(String tableNameWithType, int partitionId, DedupContext dedupContext) {
+    _tableNameWithType = tableNameWithType;
+    _partitionId = partitionId;
+    _primaryKeyColumns = dedupContext.getPrimaryKeyColumns();
+    _hashFunction = dedupContext.getHashFunction();
+    _serverMetrics = dedupContext.getServerMetrics();
+    _metadataTTL = dedupContext.getMetadataTTL() >= 0 ? dedupContext.getMetadataTTL() : 0;
+    _dedupTimeColumn = dedupContext.getDedupTimeColumn();
+    if (_metadataTTL > 0) {
+      Preconditions.checkArgument(_dedupTimeColumn != null,
+          "When metadataTTL is configured, metadata time column must be configured for dedup enabled table: %s",
+          tableNameWithType);
+    }
+    _logger = LoggerFactory.getLogger(tableNameWithType + "-" + partitionId + "-" + getClass().getSimpleName());
+  }
+
+  @Override
+  public boolean checkRecordPresentOrUpdate(PrimaryKey pk, IndexSegment indexSegment) {
+    throw new UnsupportedOperationException(
+        "checkRecordPresentOrUpdate(PrimaryKey pk, IndexSegment indexSegment) is " + "deprecated!");
+  }
+
+  @Override
+  public void addSegment(IndexSegment segment) {
+    addOrReplaceSegment(null, segment);
+  }
+
+  @Override
+  public void replaceSegment(IndexSegment oldSegment, IndexSegment newSegment) {
+    addOrReplaceSegment(oldSegment, newSegment);
+  }
+
+  private void addOrReplaceSegment(@Nullable IndexSegment oldSegment, IndexSegment newSegment) {
+    if (!startOperation()) {
```

Comment:
nit: it may be a bit more readable to move the if-check and the try-catch-finally code into addSegment/replaceSegment so that we don't need to check oldSegment==null, like the add/replace methods in upsert manager.

Reply:
good suggesion!